src/mcmh.cpp

```diff
@@ -0,0 +1,247 @@
+//' which needed to apply and reduce across a couple different
+//' dimensions.  In theory, this should also allow more optimal storage
+//' for faster calculations and less memory churn, with the downside
+//' being a more complicated function.
+//'
+//' @author Ryan Moore
+//'
+//' @param num_samples Number of samples (nrow(Y))
+//' @param Y Logratio matrix
+//' @param W Corresponding count matrix
+//' @param eY Current expected value of logratio matrix
+//' @param base OTU index used for base (1-based index)
+//' @param sigInv Output of *_network(sigma)
+//' @param mciters Number of MC iterations
+//' @param iters_to_burn How many MC iters to burn?
+//' @param stepsize Variance used for MH samples.
+//'   Tweak to adjust acceptance ratio.
+//'
+//' @return Returns a List containing Y_new and sigSum.
+// [[Rcpp::export]]
+Rcpp::List
+get_Y_new_and_sigSum(const int num_samples,
+                     const Eigen::MatrixXd& Y,
+                     const Eigen::MatrixXd& W,
+                     const Eigen::MatrixXd& eY,
+                     const int base,
+                     const Eigen::MatrixXd& sigInv,
+                     const int mciters,
+                     const int iters_to_burn,
+                     const double stepsize)
+{
+
+  const int notus = W.cols();
+  MatrixXd sigSum = MatrixXd(notus - 1, notus - 1).setZero();
+  MatrixXd Y_MH_sums = MatrixXd(notus - 1, num_samples).setZero();
+
+  // Track previous lr estimates for all samples.
+  MatrixXd previous = MatrixXd(notus - 1, num_samples).setZero();
+
+  for (int mciter = 0; mciter < mciters; ++mciter) {
+    // log ratio estimates for this sample.  Normally will have more
+    // OTUs than samples.  So save it that way to have longer bits of
+    // data in memory.
+    MatrixXd sample_lr_estimates = MatrixXd(notus - 1, num_samples);
+
+    for (int smpl = 0; smpl < num_samples; ++smpl) {
+      VectorXd Yi = Y.row(smpl);
+      VectorXd Wi = W.row(smpl);
+      VectorXd eYi = eY.row(smpl);
+
+      if (mciter == 0) {
+        previous.col(smpl) = Yi;
+      }
+
+      // Remove base from Wi
+      VectorXd Wi_no_base = remove_element(base - 1, Wi);
+
+      // Calculate estimated Yi values (Yi_star)
+      VectorXd Yi_star = make_yi_star(stepsize, Yi);
+
+      double full_ratio = get_full_ratio(Wi,
+                                         Wi_no_base,
+                                         Yi,
+                                         Yi_star,
+                                         eYi,
+                                         sigInv);
+
+      double acceptance = get_acceptance(full_ratio);
+
+      // Add estimated Yi vals to the matrix.
+      if (R::runif(0, 1) < acceptance) {
+        // accepted!
+        if (mciter >= iters_to_burn) {
+          for (int i = 0; i < notus - 1; ++i) {
+            // TODO is it off by 1?
+            Y_MH_sums(i, smpl) += Yi_star(i);
+            sample_lr_estimates(i, smpl) = Yi_star(i);
```

Comment:
So that is tracking the logratios for each mc_iter.  I'm tracking both `Y_MH_sums` and `sample_lr_estimates` because after every non-burnt mc_iter, I need the `sample_lr_estimates` matrix for a crossproduct, but I don't want to store hundreds of those matrices across the mc_iters (because you need to take the mean across iters for the Y_new value).

I do see some opportunities to clean this up a bit, fwiw.